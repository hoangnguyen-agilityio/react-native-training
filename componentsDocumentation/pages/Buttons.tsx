import React, { useState } from 'react';
import {
  StyleSheet,
  Button,
  Pressable,
  View,
  Text,
  SafeAreaView,
  Alert,
} from 'react-native';
import Section from '../components/Section';
import Title from '../components/Title';

const PressableComponent = () => {
  const [timesPressed, setTimesPressed] = useState<number>(0);

  let textLog: string = '';
  if (timesPressed > 1) {
    textLog = timesPressed + 'x onPress';
  } else if (timesPressed > 0) {
    textLog = 'onPress';
  }

  return (
    <View>
      <Pressable
        onPress={() => {
          setTimesPressed(current => current + 1);
        }}
        style={({ pressed }) => [
          {
            backgroundColor: pressed ? 'rgb(210, 230, 255)' : 'white',
          },
          styles.wrapperCustom,
        ]}>
        {({ pressed }) => (
          <Text>{pressed ? 'Pressed!' : 'Press Me'}</Text>
        )}
      </Pressable>
      <View style={styles.logBox}>
        <Text testID="pressable_press_console">{textLog}</Text>
      </View>
    </View>
  );
};

const ButtonsPage = () => (
  <SafeAreaView style={styles.container}>
    <Section>
      <Title>Default:</Title>
      <Button
        title="Press me"
        onPress={() => Alert.alert('Simple Button pressed')}
      />
    </Section>
    <Section>
      <Title>Custom color:</Title>
      <Button
        title="Press me"
        color="#f194ff"
        onPress={() => Alert.alert('Button with adjusted color pressed')}
      />
    </Section>
    <Section>
      <Title>Disabled:</Title>
      <Button
        title="Press me"
        disabled
        onPress={() => Alert.alert('Cannot press this one')}
      />
    </Section>
    <Section>
      <Title>Button width fit to content</Title>
      <View style={styles.fixToText}>
        <Button
          title="Left button"
          onPress={() => Alert.alert('Left button pressed')}
        />
        <Button
          title="Right button"
          onPress={() => Alert.alert('Right button pressed')}
        />
      </View>
    </Section>
    <Section>
      <Title>Pressable</Title>
      <PressableComponent />
    </Section>
  </SafeAreaView>
);

const styles = StyleSheet.create({
  container: {
    marginHorizontal: 16,
  },
  fixToText: {
    flexDirection: 'row',
    justifyContent: 'space-between',
  },
  wrapperCustom: {
    borderRadius: 8,
    padding: 6,
  },
  logBox: {
    padding: 20,
    margin: 10,
    borderWidth: StyleSheet.hairlineWidth,
    borderColor: '#f0f0f0',
    backgroundColor: '#f9f9f9',
  },
});

export default ButtonsPage;
