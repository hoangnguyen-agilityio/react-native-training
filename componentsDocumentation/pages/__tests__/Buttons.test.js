import React from 'react';
import renderer from 'react-test-renderer';
import Buttons from '../Buttons';
import {Button, Alert} from 'react-native';

test('renders correctly', () => {
  const tree = renderer.create(<Buttons />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Alert when click first button', () => {
  const alertSpy = jest.spyOn(Alert, 'alert');
  const tree = renderer.create(<Buttons />);

  const firstButton = tree.root.findAllByType(Button)[0];
  firstButton.props.onPress();
  expect(alertSpy).toHaveBeenCalled();
});

test('Alert when click second button', () => {
  const alertSpy = jest.spyOn(Alert, 'alert');
  const tree = renderer.create(<Buttons />);

  const secondButton = tree.root.findAllByType(Button)[1];
  secondButton.props.onPress();
  expect(alertSpy).toHaveBeenCalled();
});

test('Alert when click third button', () => {
  const alertSpy = jest.spyOn(Alert, 'alert');
  const tree = renderer.create(<Buttons />);

  const thirdButton = tree.root.findAllByType(Button)[2];
  expect(thirdButton.props.disabled).toBeTruthy();
});

test('Alert when click fourth button', () => {
  const alertSpy = jest.spyOn(Alert, 'alert');
  const tree = renderer.create(<Buttons />);

  const fourthButton = tree.root.findAllByType(Button)[3];
  fourthButton.props.onPress();
  expect(alertSpy).toHaveBeenCalled();
});

test('Alert when click fifth button', () => {
  const alertSpy = jest.spyOn(Alert, 'alert');
  const tree = renderer.create(<Buttons />);

  const fifthButton = tree.root.findAllByType(Button)[4];
  fifthButton.props.onPress();
  expect(alertSpy).toHaveBeenCalled();
});
