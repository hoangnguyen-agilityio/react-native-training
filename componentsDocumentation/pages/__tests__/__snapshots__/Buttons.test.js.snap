// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "marginHorizontal": 16,
    }
  }
>
  <View
    style={
      Object {
        "borderBottomColor": "#737373",
        "borderBottomWidth": 2,
        "marginTop": 24,
        "paddingVertical": 16,
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 18,
          "marginBottom": 12,
        }
      }
    >
      Default:
    </Text>
    <View
      accessibilityRole="button"
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "margin": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Press me
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "borderBottomColor": "#737373",
        "borderBottomWidth": 2,
        "marginTop": 24,
        "paddingVertical": 16,
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 18,
          "marginBottom": 12,
        }
      }
    >
      Custom color:
    </Text>
    <View
      accessibilityRole="button"
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "margin": 8,
                "textAlign": "center",
              },
              Object {
                "color": "#f194ff",
              },
            ]
          }
        >
          Press me
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "borderBottomColor": "#737373",
        "borderBottomWidth": 2,
        "marginTop": 24,
        "paddingVertical": 16,
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 18,
          "marginBottom": 12,
        }
      }
    >
      Disabled:
    </Text>
    <View
      accessibilityRole="button"
      accessibilityState={
        Object {
          "disabled": true,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
            Object {},
          ]
        }
      >
        <Text
          disabled={true}
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "margin": 8,
                "textAlign": "center",
              },
              Object {
                "color": "#cdcdcd",
              },
            ]
          }
        >
          Press me
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "borderBottomColor": "#737373",
        "borderBottomWidth": 2,
        "marginTop": 24,
        "paddingVertical": 16,
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 18,
          "marginBottom": 12,
        }
      }
    >
      Button width fit to content
    </Text>
    <View
      style={
        Object {
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <View
        accessibilityRole="button"
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "margin": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            Left button
          </Text>
        </View>
      </View>
      <View
        accessibilityRole="button"
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "margin": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            Right button
          </Text>
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "borderBottomColor": "#737373",
        "borderBottomWidth": 2,
        "marginTop": 24,
        "paddingVertical": 16,
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 18,
          "marginBottom": 12,
        }
      }
    >
      Pressable
    </Text>
    <View>
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "backgroundColor": "white",
            },
            Object {
              "borderRadius": 8,
              "padding": 6,
            },
          ]
        }
      >
        <Text>
          Press Me
        </Text>
      </View>
      <View
        style={
          Object {
            "backgroundColor": "#f9f9f9",
            "borderColor": "#f0f0f0",
            "borderWidth": 0.5,
            "margin": 10,
            "padding": 20,
          }
        }
      >
        <Text
          testID="pressable_press_console"
        >
          
        </Text>
      </View>
    </View>
  </View>
</RCTSafeAreaView>
`;
